### PR TITLE
Send RLPx auth in EIP-8 format

### PR DESCRIPTION
**Pull request by vorot93**
[EIP-8 specified a new handshake format for RLPx back in 2015.](https://eips.ethereum.org/EIPS/eip-8) It is far more extendable and computationally lighter on the nodes and used throughout all major clients today.

The new handshake format is also supported by OpenEthereum when another node initiates the connection. Unfortunately, however, OpenEthereum itself connects using the old handshake, most likely to support `parity` versions released prior to 2017.

I don't think this is relevant anymore. This PR upgrades outbound handshake to comply with EIP-8 as well. This is important because as I am building [rust-devp2p](https://github.com/rust-ethereum/devp2p) I have to either support two handshake formats (which requires extra maintenance and is more computationally intensive), or drop OpenEthereum to rust-devp2p connectivity. More than that, OE is the last blocker to *all* clients dropping pre-EIP-8 handshake support.

### CODE REVIEW

**Review by ordian — COMMENTED** [review]
It does make sense to me to drop the old protocol if it simplifies and optimizes thing.

But make sure it doesn't break anything by running the branch with `-l network` logs and maybe add more tests from the spec.

File: util/network-devp2p/src/handshake.rs [hunk 1]
```diff
@@ -314,27 +298,26 @@ impl Handshake {
...
-        }
-        let message = ecies::encrypt(&self.id, &[], &data)?;
-        self.auth_cipher = message.clone();
-        self.connection.send(io, message);
-        self.connection.expect(V4_ACK_PACKET_SIZE);
-        self.state = HandshakeState::ReadingAck;
+        let encoded = rlp
+            .out()
+            .into_iter()
+            .chain(repeat(0).take(100 + random::<usize>() % 200))
```
[the remark this answers is not in the document]
Comment (ordian) [reply util/network-devp2p/src/handshake.rs:311]: > range [100, 300]

if this is math notation, then it should be `rand::thread_rng().gen_range::<usize>(100, 301)`

File: util/network-devp2p/src/handshake.rs [hunk 2]
```diff
@@ -262,37 +260,23 @@ impl Handshake {
         Ok(())
     }
 
-    /// Parse and validate ack message
-    fn read_ack(&mut self, secret: &Secret, data: &[u8]) -> Result<(), Error> {
-        trace!(target: "network", "Received handshake ack from {:?}", self.connection.remote_addr_str());
-        if data.len() != V4_ACK_PACKET_SIZE {
-            debug!(target: "network", "Wrong ack packet size");
+    fn read_ack_len(&mut self, data: &[u8]) -> Result<(), Error> {
+        if data.len() != 2 {
```
[the remark this answers is not in the document]
Comment (vorot93) [reply util/network-devp2p/src/handshake.rs]: done

File: util/network-devp2p/src/handshake.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vorot93) [reply util/network-devp2p/src/handshake.rs]: done

**Comment by vorot93** [comment]
@ordian Not sure about additional tests, but connecting to ethereum-sentry works now.

File: util/network-devp2p/src/handshake.rs [hunk 3]
```diff
@@ -314,27 +311,26 @@ impl Handshake {
...
-        }
-        let message = ecies::encrypt(&self.id, &[], &data)?;
-        self.auth_cipher = message.clone();
-        self.connection.send(io, message);
-        self.connection.expect(V4_ACK_PACKET_SIZE);
-        self.state = HandshakeState::ReadingAck;
+        let encoded = rlp
+            .out()
+            .into_iter()
+            .chain(repeat_with(rand::random).take(rand::thread_rng().gen_range::<usize>(100, 301)))
```
Comment (ordian) [inline util/network-devp2p/src/handshake.rs:324]: actually looking at what geth does (although my `go` skills are weak), it seems to be zeroes in range `[100, 300)`
https://github.com/ethereum/go-ethereum/blob/97fc1c3b1d054a1345934e18fa6c3489e9119666/p2p/rlpx/rlpx.go#L592

Comment (vorot93) [reply util/network-devp2p/src/handshake.rs:324]: @ordian True, but implementations shouldn't read or touch this padding anyway.

Comment (fjl) [reply util/network-devp2p/src/handshake.rs:324]: It's OK to have random padding, but the padding will be encrypted, so it's also OK (and cheaper) to use zeros instead.

Comment (vorot93) [reply util/network-devp2p/src/handshake.rs]: Changed to zeroes

File: util/network-devp2p/src/handshake.rs [hunk 4]
```diff
@@ -314,27 +311,26 @@ impl Handshake {
...
-        }
-        let message = ecies::encrypt(&self.id, &[], &data)?;
-        self.auth_cipher = message.clone();
-        self.connection.send(io, message);
-        self.connection.expect(V4_ACK_PACKET_SIZE);
-        self.state = HandshakeState::ReadingAck;
+        let encoded = rlp
+            .out()
+            .into_iter()
+            .chain(repeat(0).take(rand::thread_rng().gen_range::<usize>(100, 301)))
```
Comment (ordian) [inline util/network-devp2p/src/handshake.rs:324]: Let's avoid premature pessimization and use a static array of zeroes and `Vec::extend` instead.
LLVM is pretty decent, but it's not magic: https://rust.godbolt.org/z/vjobKj.
The latter generates much less code and uses one single memset.

Comment (vorot93) [reply util/network-devp2p/src/handshake.rs]: Replaced with `Vec::resize`.

**Comment by rakita** [comment]
Opened new pull request with modifications here: https://github.com/openethereum/openethereum/pull/287

**Pull request closed without merging**